mfs/system.go

```diff
@@ -123,6 +123,36 @@ func (kr *Root) Flush() error {
 	return nil
 }
 
+// FlushMemFree flushes the root directory and then uncaches all of its links.
+// This has the effect of clearing out potentially stale references and allows
+// them to be garbage collected.
+// CAUTION: Take care not to ever call this while holding a reference to any
+// child directories. Those directories will be bad references and using them
+// may have unintended racy side effects.
+// A better implemented mfs system (one that does smarter internal caching and
+// refcounting) shouldnt need this method.
+func (kr *Root) FlushMemFree(ctx context.Context) error {
+	dir, ok := kr.GetValue().(*Directory)
+	if !ok {
+		return fmt.Errorf("invalid mfs structure, root should be a directory")
+	}
+
+	if err := dir.Flush(); err != nil {
+		return err
+	}
+
+	names, err := dir.ListNames(ctx)
+	if err != nil {
+		return err
+	}
+
+	for _, name := range names {
+		dir.Uncache(name)
```

Comment:
Instead of listing then uncaching, can we just uncache all cached names? That `ListNames` call could get *expensive*.

Reply:
Specifically, with sharded directories.

Reply:
Oh yeah, youre right. This will go through all names, even if they arent cached.